### ISSUE TITLE
세그먼트 모델 예시 (사진 한 장)

### ISSUE THREAD

**Issue by jkjnaver** [d0gn/Autonomous-Driving#70]
![Image](https://github.com/user-attachments/assets/6de5d80a-dfe4-404b-b335-ab43b5462c9b)

1. 이미지 불러오기 및 크기 조정
사용자가 지정한 이미지를 불러와 세그멘테이션 모델의 입력 크기(520x520)에 맞게 다시 사이즈 함.
이미지가 없으면 오류 메시지를 출력하고 종료.

2. 이미지 보정 (명암 개선)
이미지의 대비를 높이기 위해 두 가지 기법을 적용합니다:
CLAHE (적응형 히스토그램 균등화): 로컬 대비를 개선해 어두운 영역도 잘 보이도록 만듬.
히스토그램 평탄화: 전체 밝기 분포를 고르게 하여 전반적인 화질을 향상.
이 과정을 통해 딥러닝 모델이 더 정확하게 물체를 분할할 수 있도록 이미지를 정제.

3. 세그멘테이션을 위한 전처리
보정된 이미지를 PyTorch 모델 입력 형식에 맞게 변환.
RGB로 변환하고, 텐서 형태로 바꾼 뒤, 모델 학습 당시 사용한 평균과 표준편차로 정규화.

4. 딥러닝 세그멘테이션 모델 불러오기
PyTorch의 DeepLabV3+ 모델 (ResNet101 백본)을 불러옴.
이 모델은 PASCAL VOC 데이터셋 기준으로 21개의 물체 클래스를 분류가능 (예: 사람, 도로, 자동차 등).

5. 이미지 세그멘테이션 수행
전처리된 이미지를 모델에 넣고 예측을 수행합니다.
각 픽셀마다 어떤 클래스에 속하는지 예측 결과를 얻고, 클래스 번호 배열로 변환.

6. 색상 매핑
숫자로 된 클래스 결과를 사람이 보기 쉬운 색상으로 바꿈.
21개의 클래스에 각기 다른 RGB 색상을 지정하여 색으로 시각화할 수 있게 만듬.

7. 결과 시각화
원본 이미지, 세그멘테이션 결과(색상), 그리고 보정된 이미지 위에 반투명하게 결과를 오버레이한 이미지를 나란히 출력.
이를 통해 세그멘테이션이 얼마나 정확하게 이루어졌는지 직관적으로 확인 가능.

=======================================================================================

import cv2
import numpy as np
from PIL import Image
import matplotlib.pyplot as plt
import torch
from torchvision import models, transforms

image_path = 'sample.jpg'
img = cv2.imread(image_path)

img = cv2.resize(img, (520, 520))

lab = cv2.cvtColor(img, cv2.COLOR_BGR2LAB)
l, a, b = cv2.split(lab)
clahe = cv2.createCLAHE(clipLimit=3.0, tileGridSize=(8, 8))
cl = clahe.apply(l)
limg = cv2.merge((cl, a, b))
enhanced = cv2.cvtColor(limg, cv2.COLOR_LAB2BGR)

yuv = cv2.cvtColor(enhanced, cv2.COLOR_BGR2YUV)
yuv[:, :, 0] = cv2.equalizeHist(yuv[:, :, 0])
final = cv2.cvtColor(yuv, cv2.COLOR_YUV2BGR)

enhanced_rgb = cv2.cvtColor(final, cv2.COLOR_BGR2RGB)
pil_image = Image.fromarray(enhanced_rgb)

preprocess = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406],
                         std=[0.229, 0.224, 0.225]),
])
input_tensor = preprocess(pil_image).unsqueeze(0)

model = models.segmentation.deeplabv3_resnet101(pretrained=True).eval()

with torch.no_grad():
    output = model(input_tensor)['out'][0]
pred = output.argmax(0).byte().cpu().numpy()

def decode_segmap(label_mask):
    label_colors = np.array([
        [0, 0, 0], [128, 0, 0], [0, 128, 0], [128, 128, 0], [0, 0, 128],
        [128, 0, 128], [0, 128, 128], [128, 128, 128], [64, 0, 0],
        [192, 0, 0], [64, 128, 0], [192, 128, 0], [64, 0, 128],
        [192, 0, 128], [64, 128, 128], [192, 128, 128], [0, 64, 0],
        [128, 64, 0], [0, 192, 0], [128, 192, 0], [0, 64, 128],
    ])
    r = np.zeros_like(label_mask, dtype=np.uint8)
    g = np.zeros_like(label_mask, dtype=np.uint8)
    b = np.zeros_like(label_mask, dtype=np.uint8)
    for l in range(21):
        mask = label_mask == l
        r[mask] = label_colors[l, 0]
        g[mask] = label_colors[l, 1]
        b[mask] = label_colors[l, 2]
    return np.stack([r, g, b], axis=2)

seg_color = decode_segmap(pred)

original = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
overlay = cv2.addWeighted(enhanced_rgb, 0.5, seg_color, 0.5, 0)

plt.figure(figsize=(18, 6))

plt.subplot(1, 3, 1)
plt.title("Original Image")
plt.imshow(original)
plt.axis('off')

plt.subplot(1, 3, 2)
plt.title("Segmentation Result")
plt.imshow(seg_color)
plt.axis('off')

plt.subplot(1, 3, 3)
plt.title("Overlay (Enhanced + Seg)")
plt.imshow(overlay)
plt.axis('off')

plt.tight_layout()
plt.show()
